### PR TITLE
BABEL-4411 set transaction isolation cmd change

### PR DESCRIPTION
**Pull request by tanscorpio7**
### Description
Currently babelfish transactions start with session default isolation level. Users can not change this isolation level after transaction has started. On running the set transaction isolation inside a transaction, current transaction isolation does not change but all subsequent transaction will pick up this new isolation level.

Changes in this PR:

Allow changing the current transaction isolation level if no DML query has been run inside transaction.
Abort the transaction if set transaction isolation level is run after a DML query in a transaction block unless the new isolation level is same as current isolation level.
Introduce an escape hatch to ignore any invalid set transaction isolation level stmt so that transaction does not abort.

#### Engine Change:
Add check hook to guc default_transaction_isolation (all errors should be taken care here, that is assign hook should never fail)
If the conditions for setting isolation level are true returns true
If conditions are false and escape hatch is not set (strict - default value) - abort transaction with error
but if escape hatch is set to ignore for same situation - we simply ignore the set tran isolation command by setting the new isolation level to current isolation level. Which is apparently the same as ignoring the cmd.

#### Extension Change:
Introduce escape hatch for set transaction isolation 
New attribute in PLtsql_stmt_execsql structure. This attribute is needed to identify if the stmt is set transaction isolation during initialisation of internal savepoint and if set, we skip the internal save point for the current stmt.
This needs to be done since postgres does not allow setting isolation level inside a save point.

### Issues Resolved

BABEL-4411

### Check List

- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/utils/misc/guc.c [hunk 1]
```diff
@@ -12259,7 +12261,13 @@ static void
 assign_default_transaction_isolation(int newval, void *extra)
 {
 	if (guc_newval_hook)
+	{
```
Comment (surendravishnoi) [inline src/backend/utils/misc/guc.c:12264]: we should put this code into hook as well

File: src/backend/utils/misc/guc.c [hunk 2]
```diff
@@ -13075,4 +13083,51 @@ guc_set_stack_value(struct config_generic *gconf, config_var_value *val)
...
+		case XACT_READ_COMMITTED:
+			return "read committed";
+		case XACT_READ_UNCOMMITTED:
+			return "read uncommitted";
+	}
+	return "(unrecognized)";
+}
+
+static bool
+check_default_transaction_isolation(int *newval, void **extra, GucSource source)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/guc.c:13104]: any reason for having this as a separate function? can we put it new_val hook?

File: src/backend/utils/misc/guc.c [hunk 3]
```diff
@@ -13075,4 +13083,51 @@ guc_set_stack_value(struct config_generic *gconf, config_var_value *val)
...
+				ignore_isolation = GetConfigOption("babelfishpg_tsql.escape_hatch_set_transaction_isolation_level", false, false);
+
+				if(ignore_isolation && !strcmp(ignore_isolation, "ignore"))
+				{
+					*newval = XactIsoLevel;
+					return true;
+				}
+				else
+				{
+					GUC_check_errmsg("SET TRANSACTION ISOLATION failed because of setting from invalid state, set escape hatch to silently ignore");
```
Comment (surendravishnoi) [inline src/backend/utils/misc/guc.c:13124]: we should provide the name of escape hatch

File: src/backend/utils/misc/guc.c [hunk 4]
```diff
@@ -13075,4 +13083,51 @@ guc_set_stack_value(struct config_generic *gconf, config_var_value *val)
...
+
+		if ( newXactIsoLevel != XactIsoLevel && IsTransactionState() )
+		{
+			if( FirstSnapshotSet || IsSubTransaction() || (newXactIsoLevel == XACT_SERIALIZABLE && RecoveryInProgress()))
+			{
+				ignore_isolation = GetConfigOption("babelfishpg_tsql.escape_hatch_set_transaction_isolation_level", false, false);
+
+				if(ignore_isolation && !strcmp(ignore_isolation, "ignore"))
+				{
+					*newval = XactIsoLevel;
```
Comment (surendravishnoi) [inline src/backend/utils/misc/guc.c:13119]: if we put everything in a single hook, we can simply skip setting transaction isolation level

File: src/backend/utils/misc/guc.c (hunk 2 above)
Comment (tanscorpio7) [reply src/backend/utils/misc/guc.c:13104]: So having a check hook will make sure that we set both 'default_transaction_Isolation' and 'transaction_isolation' together and never just 'default_transaction_Isolation'

File: src/backend/utils/misc/guc.c (hunk 4 above)
Comment (tanscorpio7) [reply src/backend/utils/misc/guc.c:13119]: So, this was needed to make sure that default and transaction level both are always same.
We would never set just one of them 
And postgres suggests that assign hook never fails

**Pull request closed without merging**